File: LeetCode/1648. Sell Diminishing-Valued Colored Balls/solution.py
class Solution:
    def maxProfit(self, inventory: List[int], orders: int) -> int:
        inventory.sort(reverse=True)
        inventory.append(0)
        res, cnt, MOD = 0, 1, 10**9+7
        for hi, lo in zip(inventory,inventory[1:]):
            # print(hi,lo)
            if orders < (hi-lo)*cnt:
                cur = hi - orders//cnt
                res += (hi+cur+1)*(hi-cur)*cnt//2
                res += cur*(orders%cnt)
                res %= MOD
                orders = 0
                break
            if hi > lo:
                res += (hi+lo+1)*(hi-lo)*cnt//2
                res %= MOD
                orders -= (hi-lo)*cnt
            cnt += 1
        return res 


# import bisect
# class Solution:
#     def maxProfit(self, inventory: List[int], orders: int) -> int:
#         res = 0
#         MOD = 10**9+7
#         inventory.sort()
#         lo, hi = 0, 10**9
#         while lo <= hi:
#             mid = (lo+hi)//2
#             idx = bisect.bisect_left(inventory,mid)
#             cnt = sum(inventory[i]-mid for i in range(idx,len(inventory)))
#             if cnt < orders:
#                 hi = mid - 1
#             else:
#                 lo = mid + 1
#         for i, cur in enumerate(inventory):
#             if cur > lo:
#                 orders -= cur-lo
#                 res += (cur+lo+1)*(cur-lo)//2
#                 res %= MOD
#         res += orders*lo
#         res %= MOD
#         return res

File: LeetCode/1648. Sell Diminishing-Valued Colored Balls/solution.cpp
using ll = long long;

class Solution {
public:
    int maxProfit(vector<int>& inventory, int orders) {
        sort(inventory.begin(), inventory.end(), greater<>());
        inventory.push_back(0);
        ll sums = 0, pre = inventory[0];
        int MOD = 1e9+7, n = inventory.size(); 
        for (int i = 1; i < n && orders; i++){
            if (inventory[i] < pre){
                if (i*(pre-inventory[i]) <= orders){
                    sums += i*(pre+inventory[i]+1)*(pre-inventory[i])/2;
                    orders -= i*(pre-inventory[i]);
                    pre = inventory[i];
                }
                else{
                    int k = orders/i, r = orders%i;
                    sums += i*(pre+pre-k+1)*k/2;
                    pre -= k;
                    sums += r*pre;
                    orders = 0;
                }
                sums %= MOD;
                // cout << sums << " " << i <<  " " << orders << endl;
            }
        }
        return sums;

    }
};